chainindex/reconcile.go

```diff
@@ -0,0 +1,211 @@
+// It performs the following steps:
+//  1. Checks if the index is empty. If so, it returns immediately as there's nothing to reconcile.
+//  2. Finds the lowest non-reverted height in the index.
+//  3. Walks backwards from the current chain head until it finds a tipset that exists
+//     in the index and is not marked as reverted.
+//  4. Sets a boundary epoch just above this found tipset.
+//  5. Marks all tipsets above this boundary as reverted, ensuring consistency with the current chain state.
+//  6. Applies all missing un-indexed tipsets starting from the last matching tipset b/w index and canonical chain
+//     to the current chain head.
+//
+// This function is crucial for maintaining index integrity, especially after chain reorgs.
+// It ensures that the index accurately reflects the current state of the blockchain.
+func (si *SqliteIndexer) ReconcileWithChain(ctx context.Context, head *types.TipSet) error {
+	si.closeLk.RLock()
+	if si.closed {
+		si.closeLk.RUnlock()
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	if head == nil {
+		return nil
+	}
+
+	return withTx(ctx, si.db, func(tx *sql.Tx) error {
+		var hasTipset bool
+		err := tx.StmtContext(ctx, si.isTipsetMessageNonEmptyStmt).QueryRowContext(ctx).Scan(&hasTipset)
+		if err != nil {
+			return xerrors.Errorf("failed to check if tipset message is empty: %w", err)
+		}
+
+		isIndexEmpty := !hasTipset
+		if isIndexEmpty && !si.reconcileEmptyIndex {
+			log.Info("Chain index is empty and reconcileEmptyIndex is disabled; skipping reconciliation")
+			return nil
+		}
+
+		if isIndexEmpty {
+			log.Info("Chain index is empty; backfilling from head")
+			return si.backfillEmptyIndex(ctx, tx, head)
+		}
+
+		// Find the minimum applied tipset in the index; this will mark the absolute min height of the reconciliation walk
+		var reconciliationEpoch abi.ChainEpoch
+		row := tx.StmtContext(ctx, si.getMinNonRevertedHeightStmt).QueryRowContext(ctx)
+		if err := row.Scan(&reconciliationEpoch); err != nil {
+			return xerrors.Errorf("failed to scan minimum non-reverted height: %w", err)
+		}
+
+		currTs := head
+		log.Infof("Starting chain reconciliation from head height %d, reconciliationEpoch: %d", head.Height(), reconciliationEpoch)
+		var missingTipsets []*types.TipSet
+
+		// The goal here is to walk the canonical chain backwards from head until we find a matching non-reverted tipset
+		// in the db so we know where to start reconciliation from
+		// All tipsets that exist in the DB but not in the canonical chain are then marked as reverted
+		// All tpsets that exist in the canonical chain but not in the db are then applied
+
+		// we only need to walk back as far as the reconciliation epoch as all the tipsets in the index
+		// below the reconciliation epoch are already marked as reverted because the reconciliation epoch
+		// is the minimum non-reverted height in the index
+		for currTs != nil && currTs.Height() >= reconciliationEpoch {
```

Comment:
gotta care about the question above about whether there's only reverted entries in the db - if we get a `reconciliationEpoch` of `0` then we're walking the whole chain; I still think there should be limits || safeguards here

Reply:
Yeah we can introduce safeguards but I've now handled the case of index being non-empty but only containing reverted entries by setting reconciliation epoch to 0 in that case.

Reply:
@rvagg The safeguard here is the `si.maxReconcileTipsets` flag. I think we should ship good docs to ensure that it's set to a sane value as I can't think of any other extremely non-complicated way to do this given that we need to collect tipsets to apply before we before we mark everything in the index that is not in the chain as reverted and then apply the collected tipsets.